ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java

```diff
@@ -180,27 +186,52 @@ private SelectOperator translateFilterNode(final QueryFilterNode logicalNode) {
     return new SelectOperator(logicalNode, logger);
   }
 
-  private AbstractPhysicalOperator translateDataSourceNode(
-      final DataSourceNode logicalNode
-  ) {
-    boolean isTableScan = false;
+  private PullPhysicalPlanType getPlanType() {
     if (!seenSelectOperator) {
       lookupConstraints = Collections.emptyList();
-      isTableScan = true;
+      return PullPhysicalPlanType.TABLE_SCAN;
     } else if (lookupConstraints.stream().anyMatch(lc -> lc instanceof NonKeyConstraint)) {
-      isTableScan = true;
+      lookupConstraints = Collections.emptyList();
+      return PullPhysicalPlanType.TABLE_SCAN;
+    } else if (lookupConstraints.stream().allMatch(lc -> ((KeyConstraint) lc).getOperator()
+        == ConstraintOperator.EQUAL)) {
+      return PullPhysicalPlanType.KEY_LOOKUP;
+    } else if (lookupConstraints.stream().allMatch(lc -> ((KeyConstraint) lc).getOperator()
+        != ConstraintOperator.EQUAL)) {
+      return PullPhysicalPlanType.RANGE_SCAN;
+    } else {
+      //shouldn't be a case: mix of equal and range scan expressions
+      lookupConstraints = Collections.emptyList();
+      return PullPhysicalPlanType.TABLE_SCAN;
     }
+  }
+
+  private AbstractPhysicalOperator translateDataSourceNode(
+      final DataSourceNode logicalNode
+  ) {
+    pullPhysicalPlanType = getPlanType();
+    if (pullPhysicalPlanType == PullPhysicalPlanType.RANGE_SCAN
+        && !queryPlannerOptions.getRangeScansEnabled()) {
+      pullPhysicalPlanType = PullPhysicalPlanType.TABLE_SCAN;
+    }
+    if (pullPhysicalPlanType == PullPhysicalPlanType.TABLE_SCAN) {
+      if (queryPlannerOptions.getTableScansEnabled()) {
+        lookupConstraints = Collections.emptyList();
+      } else {
+        throw new KsqlException("Query requires table scan to enabled");
```

Comment:
nit: let's make the error message actionable with exactly what they should do to enable it